src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,209 @@
+        return true;
+    }
+
+    public static CreateDeviceResponse createDevice(Main main, String userId, String deviceName, int skew, int period)
+            throws StorageQueryException, DeviceAlreadyExistsException {
+
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+        String secret = generateSecret();
+
+        TOTPDevice device = new TOTPDevice(userId, deviceName, secret, skew, period, false);
+        totpStorage.createDevice(device);
+
+        // TODO: Should we just return the secret as a string?
+        return new CreateDeviceResponse(secret);
+    }
+
+    public static VerifyDeviceResponse verifyDevice(Main main, String userId, String deviceName, String code)
+            throws StorageQueryException, TotpNotEnabledException, UnknownDeviceException, InvalidTotpException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
+        // Find the device:
+        TOTPDevice matchingDevice = null;
+        for (TOTPDevice device : devices) {
+            if (device.deviceName.equals(deviceName)) {
+                if (device.verified) {
+                    // TODO: Should we just return a boolean here?
+                    return new VerifyDeviceResponse(true);
+                } else {
+                    matchingDevice = device;
+                    break;
+                }
+            }
+        }
+        if (matchingDevice == null) {
+            throw new UnknownDeviceException();
+        }
+
+        // // Insert the code into the list of used codes:
+        TOTPUsedCode newCode = new TOTPUsedCode(userId, code, true, System.currentTimeMillis() + 1000 * 60 * 5);
+        totpStorage.insertUsedCode(newCode);
+
+        // Check if the code is valid:
+        if (!checkCode(matchingDevice, code)) {
+            throw new InvalidTotpException();
+        }
+
+        // Check if the code is unused:
+        TOTPUsedCode[] usedCodes = totpStorage.getUsedCodes(userId);
+        for (TOTPUsedCode usedCode : usedCodes) {
+            if (usedCode.code.equals(code) && usedCode.isValidCode) {
+                throw new InvalidTotpException();
+            }
+        }
+
+        totpStorage.markDeviceAsVerified(userId, deviceName);
+        return new VerifyDeviceResponse(false);
+    }
+
+    public static void verifyCode(Main main, String userId, String code, boolean allowUnverifiedDevices)
+            throws StorageQueryException, TotpNotEnabledException, InvalidTotpException,
+            LimitReachedException {
+        TOTPSQLStorage totpStorage = StorageLayer.getTOTPStorage(main);
+
+        // Check if the user has any devices:
+        TOTPDevice[] devices = totpStorage.getDevices(userId);
+        if (devices.length == 0) {
+            throw new TotpNotEnabledException();
+        }
+
```

Comment:
We need to filter out unverified devices based on allowUnverifiedDevices

Reply:
Done.